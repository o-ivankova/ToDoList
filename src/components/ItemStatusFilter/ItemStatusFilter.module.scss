.itemStatusFilterPanel {
    margin-top: 10px;
    display: flex;
    flex-direction: row;
    justify-content: space-between;
}

.clearButton {
    border-style: hidden;
    background: none;
    color: #363636;
    margin-right: 5px;
    font-size: 14px;
}

.clearButton:hover {
   color: #74269b;
}