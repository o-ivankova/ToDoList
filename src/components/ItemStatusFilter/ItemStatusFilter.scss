.item-status-filter-panel {
    margin-top: 10px;
}

.btn {
    padding: 0;
    border: none;
    background: none;
}

.btn.active {
    font-weight: purple;
}

.item-status-filter-panel {
    display: flex;
    flex-direction: row;
    justify-content: space-between;
  }