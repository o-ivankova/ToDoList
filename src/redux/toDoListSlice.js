import { createSlice } from "@reduxjs/toolkit";
import storage from "redux-persist/lib/storage";
import { persistReducer } from "redux-persist";

const initialState = {
  elements: [],
  elementId: 0,
  filter: "all",
};

export const toDoListSlice = createSlice({
  name: "toDoList",
  initialState,
  reducers: {
    addElement: (state, action) => {
      const input = action.payload;

      if (input !== "") {
        state.elementId += 1;

        const newElement = {
          content: input,
          completed: false,
          id: state.elementId,
        };

        state.elements.push(newElement);
      }
    },

    deleteElement: (state, action) => {
      const id = action.payload;
      state.elements = state.elements.filter((el) => el.id !== id);
    },

    deleteAllCompleted: (state) => {
      state.elements = state.elements.filter((el) => !el.completed);
    },

    checkCompleted: (state, action) => {
      const id = action.payload;
      const index = state.elements.findIndex((el) => el.id === id);
      state.elements[index].completed = !state.elements[index].completed;
    },

    changeFilter: (state, action) => {
      const filter = action.payload;
      state.filter = filter;
    },

    reorderElements: (state, action) => {
      const result = action.payload;
      if (!result.destination) return;

      const [reorderedEl] = state.elements.splice(result.source.index, 1);
      state.elements.splice(result.destination.index, 0, reorderedEl);
    },
  },
});

export const {
  addElement,
  deleteElement,
  checkCompleted,
  deleteAllCompleted,
  changeFilter,
  reorderElements,
} = toDoListSlice.actions;

const persistConfig = {
  key: "root",
  storage,
};

const { reducer } = toDoListSlice;

export const persistedReducer = persistReducer(persistConfig, reducer);

// export default reducer;
