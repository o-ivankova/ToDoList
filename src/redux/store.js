import { configureStore } from "@reduxjs/toolkit";
import {persistedReducer} from "./toDoListSlice";
import { persistStore } from 'redux-persist'

export const store = configureStore({
  reducer: {
    toDoList: persistedReducer,
  },
});

export const persistor = persistStore(store);