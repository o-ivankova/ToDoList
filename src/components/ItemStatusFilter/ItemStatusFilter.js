import React from "react";
import "./ItemStatusFilter.scss";
import { useSelector, useDispatch } from "react-redux";
import { deleteAllCompleted, changeFilter } from "../../redux/toDoListSlice";

const ItemStatusFilter = () => {
  const elements = useSelector((state) => state.toDoList.elements);
  const filter = useSelector((state) => state.toDoList.filter);
  const dispatch = useDispatch();

  const onDeleteCompleted = () => {
    dispatch(deleteAllCompleted());
  };

  const itemsLeftCount = elements.filter((el) => !el.completed).length;

  const onFilterChange = (filter) => {
    dispatch(changeFilter(filter));
  };

  const buttonsData = [
    { name: "all", label: "All" },
    { name: "active", label: "Active" },
    { name: "completed", label: "Completed" },
  ];

  const buttons = buttonsData.map(({ name, label }) => {
    const isActive = filter === name;
    const clazz = isActive ? "active" : "";
    return (
      <button
        type="button"
        className={`btn ${clazz}`}
        key={name}
        onClick={() => onFilterChange(name)}
      >
        {label}
      </button>
    );
  });

  return (
    <span className="item-status-filter-panel">
      {itemsLeftCount} items left
      {buttons}
      <button type="btn clear" onClick={() => onDeleteCompleted()}>
        Clear completed
      </button>
    </span>
  );
};

export default ItemStatusFilter;
