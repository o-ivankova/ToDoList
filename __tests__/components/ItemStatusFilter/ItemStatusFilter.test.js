import React from "react";
import "@testing-library/jest-dom";
import { Provider } from "react-redux";
import {
  expect,
  test,
  describe,
  jest,
  afterAll,
  afterEach,
} from "@jest/globals";
import { fireEvent, render, screen } from "@testing-library/react";
import ItemStatusFilter from "../../../src/components/ItemStatusFilter";
import {setupStore} from "../../../src/redux/store";
import renderer from "react-test-renderer";

describe("FilterButtons tests", () => {
  test("component is rendered", () => {
    const preloadedState = {
      toDoList: {
        filter: "all",
        elements: [],
      },
    };

    const mockStore = setupStore(preloadedState);

    render(
      <Provider store={mockStore}>
        <ItemStatusFilter />
      </Provider>
    );

    expect(screen.getByTestId("filter-buttons")).toBeInTheDocument();
    expect(screen.getByTestId("counter")).toBeInTheDocument();
    expect(screen.getByTestId("clear-button")).toBeInTheDocument();
  });

  test("clear completed elements", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
          { id: 4, content: "test4", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);
   
    render(
      <Provider store={mockStore}>
        <ItemStatusFilter />
      </Provider>
    );

    fireEvent.click(screen.getByTestId("clear-button"));

    const expectedLocalStorage = preloadedState.toDoList.elements.filter(
      (el) => !el.completed
    );
    const elementsInLocalStorage = JSON.parse(
      localStorage.getItem("toDoElements")
    );
    
    expect(elementsInLocalStorage).toEqual(expectedLocalStorage);
  });

  test("clear completed elements - no completed elements", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: false },
          { id: 2, content: "test2", completed: false },
          { id: 3, content: "test3", completed: false },
          { id: 4, content: "test4", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);
   
    render(
      <Provider store={mockStore}>
        <ItemStatusFilter />
      </Provider>
    );

    fireEvent.click(screen.getByTestId("clear-button"));

    const expectedLocalStorage = preloadedState.toDoList.elements;

    const elementsInLocalStorage = JSON.parse(
      localStorage.getItem("toDoElements")
    );
    
    expect(elementsInLocalStorage).toEqual(expectedLocalStorage);
  });

  afterAll(() => {
    jest.clearAllMocks();
  });

  test("snapshot test", () => {
    const preloadedState = {
      toDoList: {
        elements: [
          { id: 1, content: "test1", completed: true },
          { id: 2, content: "test2", completed: true },
          { id: 3, content: "test3", completed: false },
        ],
      },
    };
    const mockStore = setupStore(preloadedState);

    const domTree = renderer
      .create(
        <Provider store={mockStore}>
          <ItemStatusFilter />
        </Provider>
      )
      .toJSON();
    expect(domTree).toMatchSnapshot();
  });

  afterEach(() => {
    localStorage.setItem("toDoElements", JSON.stringify([]));
  });
});
