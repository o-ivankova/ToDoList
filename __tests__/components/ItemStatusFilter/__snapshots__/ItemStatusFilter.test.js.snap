// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterButtons tests snapshot test 1`] = `
<span
  className="itemStatusFilterPanel"
  data-testid="status-filter-panel"
>
  <div
    className="counter"
    data-testid="counter"
  >
    1
     items left
  </div>
  <div
    data-testid="filter-buttons"
  >
    <button
      className="filterButton"
      onClick={[Function]}
      type="button"
    >
      All
    </button>
    <button
      className="filterButton"
      onClick={[Function]}
      type="button"
    >
      Active
    </button>
    <button
      className="filterButton"
      onClick={[Function]}
      type="button"
    >
      Completed
    </button>
  </div>
  <button
    className="clearButton"
    data-testid="clear-button"
    onClick={[Function]}
  >
    Clear completed
  </button>
</span>
`;
